File: Arrays & Hashing/encode_and_decode_strings.js
/*
https://leetcode.com/problems/encode-and-decode-strings/description/
*/

class Solution { 
    constructor(){
        this.hashmap = new Map();
        this.pointer = 0;
    }

    /**
     * @param {string[]} strs
     * @returns {string}
     #Time complexity O(n), Space complexity O(n)
     */
    encode(strs) {
        var res = ""

        for(let s of strs){
            res += String(this.pointer)
            res += ","
            this.hashmap.set(this.pointer, s)
            this.pointer += 1
        }     

        return res
    }

    /**
     * @param {string} str
     * @returns {string[]}
     */
    decode(str) {
        const res = []

        for(let c of str.split(",").slice(0, -1)){
            res.push(this.hashmap.get(Number(c)))
        }

        return res
    }
}

File: Arrays & Hashing/encode_and_decode_strings.py
"""
https://leetcode.com/problems/encode-and-decode-strings/description/
"""

class Solution: #Time complexity O(n), Space complexity O(n)

    def __init__(self):
        self.hashmap = {}
        self.pointer = 0

    def encode(self, strs: List[str]) -> str:
        res = ""
        for s in strs:
            res += str(self.pointer)
            res += ","
            self.hashmap[self.pointer] = s
            self.pointer += 1

        return res

    def decode(self, s: str) -> List[str]:
        res = []
        for c in s.split(",")[0:-1]:
            res.append(self.hashmap[int(c)])

        return res

